### ISSUE TITLE
Debug Support for Any Exceptions and Uncaught Exceptions

### ISSUE THREAD

**Issue by sgtoj** [PowerShell/vscode-powershell#298]
Is it possible to debug support to enable break points for: Any Exceptions, or Uncaught Exceptions?

![VS Code Debug Breakpoints](http://i.imgur.com/MKna0uD.png)

**Comment by daviwil** [comment]
I'm not sure how this would work in the case of PowerShell.  What do you think, @rkeithhill?

**Comment by rkeithhill** [comment]
I've wanted this for a while but I think PowerShell would need to be modified to support this.  Who's the debugger expert on the team? Paul H?

FYI, I filed a UserVoice suggestion on this back in March - please vote it up: https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/12843021-enhance-set-psbreakpoint-to-allow-us-to-set-except

**Comment by daviwil** [comment]
Yep, @PaulHigin is the PowerShell debugging expert.

**Comment by PaulHigin** [comment]
This is a fairly common request and something we have in our backlog, but is impossible to implement with our current script debugger.  The reason is that the script debugger runs on the script execution thread and requires running script to work.  But once an exception is thrown on the script execution thread the script is no longer running and the script debugger no longer works.  At this point you need a managed/native debugger.   We could have script debugger break for non-terminating errors since script execution continues and the script debugger is working.  But I am not sure how useful that would be.

Jason, @lzybkr, has experimented with a mixed managed code/script debugger and it would be ideal for this kind of thing.  When the managed exception is thrown this debugger could show both managed and PowerShell script stacks.

**Comment by daviwil** [comment]
The mixed code/script debugger would be extremely nice to have and would be pretty easy to surface in VS Code.

**Comment by lzybkr** [comment]
That prototype generated a pdb and code that the C# debugger could consume - it was useful, but not a PowerShell experience.

**Comment by rkeithhill** [comment]
@PaulHigin What do you think about this workaround by @nightroman?  https://github.com/nightroman/PowerShelf/blob/master/Debug-Error.ps1

**Comment by PaulHigin** [comment]
Very clever.  I like it.  But I believe this just handles the case where PowerShell is throwing the exception (either through throw keyword or -ErrorAction Stop).  @lzybkr can you confirm?  I was thinking in terms of a general exception being thrown in managed code.  Still this looks to be very useful and something we could support internally.

**Comment by nightroman** [comment]
Unfortunately, `Debug-Error.ps1` does not show the error itself, it just breaks into the debugger. The error has not been written yet (i.e. added to $Error) by PowerShell. The tool is useful only in special cases, rather rare.

**Comment by andrewducker** [comment]
Any movement on this?  At the moment I'm reduced to stepping through my module (and all of its dependencies) line by line trying to find where the exception is being thrown.  Or adding a load of Write-Host statements to see which ones get hit.

**Comment by SeeminglyScience** [comment]
There is a way to break before the terminating error is thrown, but after it's been written to `$Error`. It isn't very clean though.

```powershell
Set-PSBreakpoint -Variable StackTrace -Mode Write -Action {
    $null = Set-PSBreakpoint -Variable ErrorActionPreference -Mode Read
}

throw 'test'
# In debugger:
$Error[0]
# Returns:
# test
# At line:1 char:1
# + throw 'test'
# + ~~~~~~~~~~~~
#     + CategoryInfo          : OperationStopped: (test:String) [], RuntimeException
#     + FullyQualifiedErrorId : test
```

After `$StackTrace` is written to, `$ErrorActionPreference` is checked to see if the error should be terminating.  When it checks `$ErrorActionPreference` it has already written to `$Error` regardless of the preference value.

For this to be reliable you would need to store state between the two breakpoints to ensure they fire on the same sequence point (in case the script later reads the preference directly).  It would also need to handle situations where the break point already exists and be able to remove itself after the first hit.

I'll start looking into the feasibility of adding support for this.

**Comment by JustinGrote** [comment]
To follow up on @SeeminglyScience's work:

- Capturing $Error by breaking on erroractionpreference only works reliably on the "throw" keyword, other exceptions don't seem to call it.
- The breakpoints don't get cleaned up after a session and can trigger again if you look at erroractionpreference, etc. so I'm looking into having a remove-breakpoint somewhere maybe, via delayed runspace or something, because you cant remove the breakpoint within itself as far as I can tell.
- Foreach Loops get weird, some additional workarounds added there
- The only way to capture $Error for non-throw items as far as I've found is to try-catch the failed line again to capture the error. This obviously can have issues if the command wasn't idempotent so I made it a toggleable variable. 
- You can't break on a write to $Error because its marked as a readonly constant, so it doesn't technically "change" from the perspective of the breakpoint watcher even though PS updates it in the background.

**Effective Workaround**
```powershell
$null = Set-PSBreakpoint -Variable stacktrace -Mode Write -Action {
    write-host [Environment]::NewLine
    write-host -fore red "Exception Happened at '$($myinvocation.line)'"

    #Skip some non-exception related stacktrace updates
    if ($stacktrace -match 'CheckActionPreference|ThrowTerminatingError|ForEachObjectCommand\.ProcessRecord') {
        write-host -fore DarkYellow "Skipping this stacktrace"
        continue
    }

    #If this error was from a throw statement, wait for throw to read erroractionpreference, at which point $Error will be updated 
    #Ref: https://github.com/PowerShell/vscode-powershell/issues/298#issuecomment-343485564
    if ([Management.Automation.Language.Parser]::ParseInput(
        $MyInvocation.Line,[ref]$null,[ref]$null).FindAll(
            {$args[0] -is [System.Management.Automation.Language.ThrowStatementAst]},$true
        )
    ) {
        if (-not (Set-PSBreakpoint -Variable ErrorActionPreference -Mode Read)) {
            $null = Set-PSBreakpoint -Variable ErrorActionPreference -Mode Read
        }
    } else {
        if (-not $PSDebugDontAttemptErrorCapture) {
            
            write-host -fore darkyellow "Re-Executing '$($myinvocation.line)' to attempt to capture the error. You can disable this behavior by setting `$PSDebugDontAttemptErrorCapture to `$false"
            try {
                $null = Invoke-Expression $myinvocation.line
            } catch {
                $GLOBAL:PSDebugCapturedError = $PSItem
            }
            if ($GLOBAL:PSDebugCapturedError) {
                Write-Host -fore DarkYellow "Error capture succeeded and stored to `$PSDebugCapturedError"
            }
        }
        break
    }
}
```

**Comment by JustinGrote** [comment]
OK, I've come up with a pretty effective function to break on exceptions in lieu of a proper implementation (https://github.com/PowerShell/PowerShell/issues/2830)

## Features

1. Breaks on any terminating exception (doesn't catch non-terminating errors on purpose unless erroractionpreference='stop' is set)
1. Re-Execution in order to capture the relevant exception (toggleable behavior)
1. Cleans up after itself (usually)
1. WinPS5 and PS6+ compatible
1. Works pretty much anywhere Powershell does (tested: vscode, cloud shell, azure functions, pwsh, etc.)

## Example launch.json entry
```json
{
    "name": "PowerShell Trap Exceptions",
    "type": "PowerShell",
    "request": "launch",
    "script": "Debug-OnException",
    "args": [
        "${file}"
    ],
    "cwd": "${file}"
}
```
## Demos
### VSCode Unsaved File
![DebugExceptionUnsavedFile](https://user-images.githubusercontent.com/15258962/61674409-95d8b000-aca8-11e9-8e1d-8997a59d8113.gif)
### VSCode Run Selection (F8)
![DebugExceptionF8](https://user-images.githubusercontent.com/15258962/61674436-b43eab80-aca8-11e9-8b45-ea0011ee5d68.gif)
### Foreach Loop
![DebugExceptionForeach](https://user-images.githubusercontent.com/15258962/61674452-c6204e80-aca8-11e9-92cf-6256fb360ab0.gif)
### Foreach (More Real-World Example)
![DebugExceptionForEachRealWorld](https://user-images.githubusercontent.com/15258962/61674495-f2d46600-aca8-11e9-9b88-51289dec73f9.gif)
### Exception Inside Module
![DebugOnModuleException](https://user-images.githubusercontent.com/15258962/61674506-0253af00-aca9-11e9-864b-867b59e58c21.gif)
### Azure Cloud Shell
![DebugCloudShellException](https://user-images.githubusercontent.com/15258962/61674703-cf5deb00-aca9-11e9-8aab-dc1ba882db56.gif)
### Powershell 5 (Windows Terminal, Works in VSCode too)
![PS5Debug](https://user-images.githubusercontent.com/15258962/61674682-bead7500-aca9-11e9-9601-ed033be43ab6.gif)

## VSCode Extension Integration

Will need help here as I suck at typescript. The script itself or the breakpoints it generates need to be wired up to a custom breakpoint "Terminating Exceptions" per @sgtoj's example. @TylerLeonhardt @SeeminglyScience @rjmholt  maybe?

**Comment by SydneyhSmith** [comment]
@JustinGrote thanks for all your work here, this looks great. The change here will need to happen in Editor Services, probably in [this chunk of code ](https://github.com/PowerShell/PowerShellEditorServices/blob/55d01fc5f6e940c5e4f2ef7bcc68636a45a94e55/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs#L677-L704), using this https://microsoft.github.io/debug-adapter-protocol/specification#Requests_SetExceptionBreakpoints ....it would be awesome if you opened up a PR in that repo and we can help you iron out the details...Thanks!

**Comment by bormm** [comment]
The script @JustinGrote created did not helps if the code itself also throws exceptions that are catched by intention. The script terminates in that case, where it would continue without the script.
I had an other solution: I just added a 
```
trap { 
# whatever
}
```
and placed a breakpoint in that trap. This worked perfectly for me.

**Comment by JustinGrote** [comment]
There is also a PR to PowerShell 7 to specifically add this functionality to set-breakpoint, so it will be easier to support going forward.

**Comment by rkeithhill** [comment]
Three years later ...  But hey, I'll be happy to finally have this support in the PS engine.  :-)

**Comment by KirkMunro** [comment]
Not yet.

I break into the debugger the moment the exception is raised.

I'll have to think about how that could be done. The only thing that comes to mind at the moment is reverse traversal of the AST to check for traps or catch blocks, but figuring out what handles what via inspection could be pretty complicated.

**Comment by SeeminglyScience** [comment]
@KirkMunro 

I think a exception could be considered unhandled when it reaches [this code block in `LocalPipeline`](https://github.com/PowerShell/PowerShell/blob/134cb5949ad08807993dea094ddb3dc61db84249/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs#L612-L617) and maybe when `IsNested` is `false`? Not sure about the latter, maybe it's just enough to hit that block.

**Comment by KirkMunro** [comment]
@SeeminglyScience: Even if that would work, I'm also concerned about the user experience. Today `ActionPreference.Break` enters the debugger the moment the corresponding exception or stream message occurs. The debugger is on the very line where the exception came from, allowing users to inspect the environment at that point in time, whether the exception is handled or not.

If we defer entering the debugger for unhandled exceptions until such a time that we know the exception is considered unhandled, where are we in our call stack? We may have unwound quite a bit to get to that point. IMHO it is far better for this community to bring users to the origin of the exception, so that they can understand what is going on.

That said, I'm all for efficient debugging (it's something I continue to actively work on in many areas), so I'd like to offer both, but I would want to do so with the debugger still breaking at the point of exception. We could do more work in the script block compiler so that we know the explicit types of exceptions that can be caught when an exception is first raised, to identify whether or not that exception is going to be handled, but then I'm not sure how to reliably identify if a catch all will handle an exception or not.

**Comment by SeeminglyScience** [comment]
> If we defer entering the debugger for unhandled exceptions until such a time that we know the exception is considered unhandled, where are we in our call stack? We may have unwound quite a bit to get to that point.

Maybe, but the snippet I linked is at the end of a very long chain of try/catch's.  With some experimentation you can probably find one where the stack is mostly intact and you can still determine if it's unhandled.

**Comment by JustinGrote** [comment]
Now that Powershell 7 is out, perhaps it's as "easy" as just wiring up the "Uncaught Exceptions" button to set ErrorActionPreference = 'break' for now?
![image](https://user-images.githubusercontent.com/15258962/76112238-acf9b300-5f96-11ea-9d85-20b6cb2ec1aa.png)

@SeeminglyScience @KirkMunro @rkeithhill 

I'll be able to take a stab at a PR in a couple weeks, knowing basically no Typescript and just enough C# to read it and interpret it into powershell :).

**Comment by rjmholt** [comment]
> knowing basically no Typescript and just enough C# to read it and interpret it into powershell 

Very happy to help with that. The hard won't be the language as much as navigating the code, but I think we have a fair idea of where things should go.

**Comment by rjmholt** [comment]
The other hard part is managing the version-specificity. It's a pain writing code that works in 7 but fails gracefully and informatively in older PowerShells.

**Comment by SeeminglyScience** [comment]
@KirkMunro do you know if there's a way to only break on uncaught errors? Or otherwise determine that an error was uncaught?  Also determine whether the error was terminating?

Right now it's all errors, even if caught.  That's still awesome and way better than no error breakpoints period though.

**Comment by JustinGrote** [comment]
@SeeminglyScience You're right! For some reason I thought it was working on uncaught exceptions only.
![image](https://user-images.githubusercontent.com/15258962/76126667-8f3b4680-5fb4-11ea-9374-78364fddcfb1.png)

Sounsd like more of a upstream (Powershell) bug to me, since ErrorActionPreference = 'stop' doesn't stop on uncaught exceptions.

**Comment by SeeminglyScience** [comment]
@JustinGrote Yeah for sure, there's still great value in enabling "All Exceptions".

**Comment by rjmholt** [comment]
> since ErrorActionPreference = 'stop' doesn't stop on uncaught exceptions so the behavior is inconsistent

Do you mean `$ErrorActionPreference = 'Stop'` doesn't stop on *caught* exceptions?

If so, what really happens is that it does "Stop" (by which I mean throw a terminating error), but then that error is caught.

To change the behaviour of `$ErrorActionPreference = 'Debug'` I feel that's more of a consistency break, since unlike other error actions it would need to search up the stack to determine if it's going to be caught or not. Honestly that's a pretty complex behaviour.

Anyway, didn't mean to sidetrack you.

**Comment by SeeminglyScience** [comment]
> since unlike other error actions it would need to search up the stack to determine if it's going to be caught or not. Honestly that's a pretty complex behaviour.

My thought here is to add something to the compiler to emit something when processing a `TryStatementAst` that would allow tracking this easier.  Still pretty complex but would be incredibly useful.

**Comment by markm77** [comment]
[Aside]

> Jason, @lzybkr, has experimented with a mixed managed code/script debugger and it would be ideal for this kind of thing. When the managed exception is thrown this debugger could show both managed and PowerShell script stacks.

I noticed this comment because I'm looking for a way to step through C# in my own PS module (dll) called by PowerShell script being debugged in vscode-powershell. At the moment I can't debug the PowerShell and C# together so @lzybkr if you ever chose to make available something that supports this I'd be grateful for one. 😊

**Comment by TylerLeonhardt** [comment]
@markm77 you are able to have the PowerShell debugger and C# debugger running at the same time.... so feasibly you can launch debug a PowerShell script and attach the C# debugger... but it doesn't give you the ideal behavior of "stepping into C# from PowerShell".

**Comment by lzybkr** [comment]
@markm77 - mixed debugging is a little painful but possible today if you attach a C# debugger, set breakpoints where you care about, then debug the PowerShell code with the PowerShell debugger. You don't get a nice shared stack but it's better than nothing.

My prototypes were based on the Desktop version of .Net, I have no idea if they are even possible now.

**Comment by markm77** [comment]
Thanks guys for interesting comments. To do mixed debugging do I have to re-attach the debugger *each time* the PowerShell script calls into a C# module cmdlet? Or can I attach the debugger just once following module import and it will re-attach as the DLL is called (much better)?

Also is there any recommended way for dropping the lock on the DLL to allow DLL re-build following any C# update? At the moment I run pwsh inside PowerShell before module import, then run exit to drop the lock on the DLL before C# re-build.

Cheers!

**Comment by lzybkr** [comment]
Attaching once is fine, but you must exit the process to rebuild - this is a limitation in .Net.

**Comment by markm77** [comment]
Thanks! I think I'm going to experiment with an outer script that does the import otherwise set up VS Code tasks. Have a great day!

**Comment by SetTrend** [comment]
I'm not sure if I get the current situation right …

According to [https://github.com/PowerShell/PowerShell/issues/2830](https://github.com/PowerShell/PowerShell/issues/2830#issuecomment-697025027), PowerShell seems to now support breaking into the debugger on uncaught exceptions. Yet, I can't find this working in the current version of VS Code PowerShell extension.

**Comment by JustinGrote** [comment]
@SetTrend I don't know if we've wired up the button specifically but if you do `$ErrorActionPreference = 'break'` it should still break in debugging when an error occurs.

@SeeminglyScience has a breakpoint PR we are waiting that changes a lot of things that we are waiting on before we make any breakpoint feature changes.
https://github.com/PowerShell/vscode-powershell/pull/4065
